iiblib/iib_build_details_model.py

```diff
@@ -1,135 +1,278 @@
+    ]
+
+    _general_attrs = [
+        "id",
+        "arches",
+        "state",
+        "state_reason",
+        "request_type",
+        "batch",
+        "updated",
+        "user",
+    ]
+
+    _optional_attrs = {
+        "state_history": lambda: list(),
+        "batch_annotations": lambda: dict(),
+        "logs": lambda: dict(),
+    }
+
+    _operation_attrs = []
+
+    def __init__(self, *args, **kwargs):
+        self._data = self._get_args(kwargs)
 
     @classmethod
     def from_dict(cls, data):
-        return cls(
-            data["id"],
-            data["state"],
-            data["state_reason"],
-            data.get("state_history", []),
-            data["from_index"],
-            data["from_index_resolved"],
-            data.get("bundles", []),
-            data.get("removed_operators", []),
-            data.get("organization"),
-            data["binary_image"],
-            data["binary_image_resolved"],
-            data["index_image"],
-            data["request_type"],
-            data["arches"],
-            data["bundle_mapping"],
-            data.get("omps_operator_version", {}),
-        )
+        if data["request_type"] == "add":
+            return AddModel(**data)
+        elif data["request_type"] == "rm":
+            return RmModel(**data)
+        elif data["request_type"] == "regenerate-bundle":
+            return RegenerateBundleModel(**data)
+        raise KeyError("Unsupported request type: %s" % data["request_type"])
+
+    def _get_args(self, data):
+        attrs = {}
+        for general_attr in self._general_attrs:
+            attrs[general_attr] = data[general_attr]
+        for optional_attr, default_maker in self._optional_attrs.items():
+            if optional_attr in data:
+                attrs[optional_attr] = data[optional_attr]
+            else:
+                attrs[optional_attr] = default_maker()
+        for operation_attr in self._operation_attrs:
+            attrs[operation_attr] = data[operation_attr]
+        return attrs
 
     def to_dict(self):
-        return {
-            "id": self.id,
-            "state": self.state,
-            "state_reason": self.reason,
-            "state_history": self.state_history,
-            "from_index": self.from_index,
-            "from_index_resolved": self.from_index,
-            "bundles": self.bundles,
-            "removed_operators": self.removed_operators,
-            "organization": self.organization,
-            "binary_image": self.binary_image,
-            "binary_image_resolved": self.binary_image_resolved,
-            "index_image": self.index_image,
-            "request_type": self.request_type,
-            "arches": self.arches,
-            "bundle_mapping": self.bundle_mapping,
-            "omps_operator_version": self.omps_operator_version,
-        }
+        return self._data
 
     def __eq__(self, other):
+        return isinstance(other, self.__class__) and self._data == other._data
```

Comment:
I am just thinking out loud, since we are implementing this (and `__getattribute__`)  by our selves, would be good idea to test is as well? Try to compare two same classes and different ones (AddModel x AddModel - should pass, AddModel x RmModel - should fail, ...) Or is it pointless?

Reply:
The __eq__ is in the tests line #243 test `assert model1 == add_model`

Reply:
and I add a test AddModel != RmModel line #294

Reply:
I am blind, thanks